File: scripts/optimize-media.py
#!/usr/bin/env python3
"""Convert heavy GIF/PNG assets to WebP posters + MP4 loops for web delivery."""
from __future__ import annotations

import json
import re
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
PUBLIC = ROOT / "public"
MAX_WIDTH = 808
WEBP_QUALITY = 82
MP4_CRF = 28


def max_width_for(path: Path) -> int:
    name = path.name
    if "_rw_3840" in name:
        return 3840
    if "_rw_1920" in name:
        return 1920
    if "_rw_1200" in name:
        return 1200
    if "_rw_600" in name:
        return 600
    if path.parent.name == "covers":
        return MAX_WIDTH
    return 1920

GIF_COVERS = [
    "atendimento-mateus.gif",
    "cubo.gif",
    "dark-feelings-2d-animation.gif",
    "the-life-of-ratildo-flash-animation.gif",
    "unused-2d-animation.gif",
]

PNG_COVERS = [
    "olhe-e-aprenda.png",
    "copia-de-smart-financeiro.png",
    "smart-financeiro.png",
    "tutti-frutti-visual-identity.png",
]


def run(cmd: list[str]) -> None:
    result = subprocess.run(
        cmd, check=False, stdout=subprocess.DEVNULL, stderr=subprocess.PIPE
    )
    if result.returncode != 0:
        raise subprocess.CalledProcessError(
            result.returncode, cmd, stderr=result.stderr
        )


def gif_to_webp_and_mp4(src: Path) -> tuple[Path, Path | None]:
    webp = src.with_suffix(".webp")
    mp4 = src.with_suffix(".mp4")
    frame = src.with_suffix(".frame.png")
    width = max_width_for(src)
    scale = f"scale='min({width},iw)':-2"

    run(
        [
            "ffmpeg",
            "-y",
            "-i",
            str(src),
            "-vf",
            scale,
            "-frames:v",
            "1",
            str(frame),
        ]
    )
    run(["cwebp", "-q", str(WEBP_QUALITY), str(frame), "-o", str(webp)])
    frame.unlink(missing_ok=True)

    try:
        run(
            [
                "ffmpeg",
                "-y",
                "-i",
                str(src),
                "-vf",
                scale,
                "-c:v",
                "libx264",
                "-pix_fmt",
                "yuv420p",
                "-movflags",
                "+faststart",
                "-an",
                "-crf",
                str(MP4_CRF),
                str(mp4),
            ]
        )
        return webp, mp4
    except subprocess.CalledProcessError:
        mp4.unlink(missing_ok=True)
        try:
            run(
                [
                    "ffmpeg",
                    "-y",
                    "-i",
                    str(src),
                    "-vf",
                    f"fps=15,{scale},split[s0][s1];[s0]palettegen[p];[s1][p]paletteuse",
                    "-c:v",
                    "libx264",
                    "-pix_fmt",
                    "yuv420p",
                    "-movflags",
                    "+faststart",
                    "-an",
                    "-crf",
                    str(MP4_CRF),
                    str(mp4),
                ]
            )
            return webp, mp4
        except subprocess.CalledProcessError:
            mp4.unlink(missing_ok=True)
            return webp, None


def png_to_webp(src: Path) -> Path:
    webp = src.with_suffix(".webp")
    frame = src.with_suffix(".frame.png")
    run(
        [
            "ffmpeg",
            "-y",
            "-i",
            str(src),
            "-vf",
            f"scale='min({MAX_WIDTH},iw)':-2",
            "-frames:v",
            "1",
            str(frame),
        ]
    )
    run(["cwebp", "-q", str(WEBP_QUALITY), str(frame), "-o", str(webp)])
    frame.unlink(missing_ok=True)
    return webp


def compress_mp4(src: Path) -> None:
    tmp = src.with_suffix(".tmp.mp4")
    run(
        [
            "ffmpeg",
            "-y",
            "-i",
            str(src),
            "-c:v",
            "libx264",
            "-crf",
            str(MP4_CRF),
            "-preset",
            "slow",
            "-movflags",
            "+faststart",
            "-an",
            str(tmp),
        ]
    )
    tmp.replace(src)


def human_size(path: Path) -> str:
    size = path.stat().st_size
    for unit in ("B", "KB", "MB"):
        if size < 1024 or unit == "MB":
            return f"{size / (1024 ** (0 if unit == 'B' else (1 if unit == 'KB' else 2))):.1f} {unit}" if unit != "B" else f"{size} B"
        size /= 1024
    return f"{size:.1f} MB"


def replace_gif_paths_in_json(root: Path) -> int:
    pattern = re.compile(r"(\/images\/[^\"']+?)\.gif")
    count = 0

    for json_path in root.rglob("*.json"):
        text = json_path.read_text(encoding="utf-8")
        if ".gif" not in text:
            continue
        updated = pattern.sub(r"\1.webp", text)
        if updated != text:
            json_path.write_text(updated, encoding="utf-8")
            count += 1

    return count


def convert_asset_gifs(assets_dir: Path) -> None:
    for gif in sorted(assets_dir.glob("*.gif")):
        webp = gif.with_suffix(".webp")
        mp4 = gif.with_suffix(".mp4")
        if webp.exists() and (mp4.exists() or not gif_to_mp4_candidate(gif)):
            continue
        if webp.exists() and not mp4.exists():
            try:
                width = max_width_for(gif)
                scale = f"scale='min({width},iw)':-2"
                run(
                    [
                        "ffmpeg",
                        "-y",
                        "-i",
                        str(gif),
                        "-vf",
                        scale,
                        "-c:v",
                        "libx264",
                        "-pix_fmt",
                        "yuv420p",
                        "-movflags",
                        "+faststart",
                        "-an",
                        "-crf",
                        str(MP4_CRF),
                        str(mp4),
                    ]
                )
                continue
            except subprocess.CalledProcessError:
                pass
        print(f"  asset {gif.name} …")
        gif_to_webp_and_mp4(gif)


def gif_to_mp4_candidate(gif: Path) -> bool:
    return gif.stat().st_size > 120_000


def main() -> int:
    covers_dir = PUBLIC / "images" / "covers"
    site_dir = PUBLIC / "images" / "site"
    videos_dir = PUBLIC / "videos"
    assets_dir = PUBLIC / "images" / "assets"

    print("Cover GIFs → WebP + MP4")
    for name in GIF_COVERS:
        src = covers_dir / name
        if not src.exists():
            print(f"  skip missing {name}")
            continue
        before = src.stat().st_size
        webp, mp4 = gif_to_webp_and_mp4(src)
        mp4_size = mp4.stat().st_size // 1024 if mp4 else 0
        print(
            f"  {name}: {before // 1024}KB → webp {webp.stat().st_size // 1024}KB"
            + (f", mp4 {mp4_size}KB" if mp4 else " (static webp only)")
        )

    print("Cover PNGs → WebP")
    for name in PNG_COVERS:
        src = covers_dir / name
        if not src.exists():
            continue
        before = src.stat().st_size
        webp = png_to_webp(src)
        print(f"  {name}: {before // 1024}KB → webp {webp.stat().st_size // 1024}KB")

    og_gif = site_dir / "og-image.gif"
    if og_gif.exists():
        og_webp = site_dir / "og-image.webp"
        og_jpg = site_dir / "og-image.jpg"
        og_frame = site_dir / "og-image.frame.png"
        run(
            [
                "ffmpeg",
                "-y",
                "-i",
                str(og_gif),
                "-vf",
                "scale=1200:-2",
                "-frames:v",
                "1",
                str(og_frame),
            ]
        )
        run(["cwebp", "-q", str(WEBP_QUALITY), str(og_frame), "-o", str(og_webp)])
        run(
            [
                "ffmpeg",
                "-y",
                "-i",
                str(og_frame),
                "-frames:v",
                "1",
                str(og_jpg),
            ]
        )
        og_frame.unlink(missing_ok=True)
        print(f"  og-image: gif {og_gif.stat().st_size // 1024}KB → webp/jpg")

    cowboys = videos_dir / "cowboys-vs-cyborgs-game-art.mp4"
    if cowboys.exists():
        before = cowboys.stat().st_size
        compress_mp4(cowboys)
        print(f"  cowboys mp4: {before // 1024}KB → {cowboys.stat().st_size // 1024}KB")

    print("Project asset GIFs (full-size only) …")
    convert_asset_gifs(assets_dir)

    print("Updating JSON paths .gif → .webp …")
    updated = replace_gif_paths_in_json(ROOT / "src" / "content")
    print(f"  updated {updated} JSON files")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/optimize-media.mjs
#!/usr/bin/env node
/**
 * Optimize all project media with sharp (images) + ffmpeg (GIF→MP4).
 *
 * What it does:
 * 1. Convert cover GIFs → WebP (poster) + MP4 (loop)
 * 2. Convert cover PNGs/JPGs → WebP
 * 3. For every asset image: generate WebP at 600/1200/1920 widths
 * 4. Remove oversized variants (_rw_3840) and placeholder files (176B)
 * 5. Update all JSON references: .gif→.webp, .png→.webp
 * 6. Compress the single project video
 */

import { execSync } from "child_process";
import { readFileSync, readdirSync, writeFileSync, existsSync, unlinkSync } from "fs";
import { readdir, rename as renameAsync, stat, unlink } from "fs/promises";
import path from "path";
import sharp from "sharp";

const ROOT = decodeURIComponent(new URL("..", import.meta.url).pathname);
const PUBLIC = path.join(ROOT, "public");
const COVERS = path.join(PUBLIC, "images", "covers");
const ASSETS = path.join(PUBLIC, "images", "assets");
const SITE = path.join(PUBLIC, "images", "site");
const VIDEOS = path.join(PUBLIC, "videos");
const CONTENT = path.join(ROOT, "src", "content");

const WEBP_QUALITY = 85;
const COVER_MAX_W = 808;
const ASSET_WIDTHS = [600, 1200, 1920];
const PLACEHOLDER_SIZE = 200; // bytes — files smaller than this are broken exports

function log(...args) {
  console.log(`[optimize]`, ...args);
}

function run(cmd) {
  execSync(cmd, { stdio: "ignore" });
}

function humanSize(bytes) {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

function extname(file) {
  return path.extname(file).toLowerCase();
}

// ─── COVERS ────────────────────────────────────────────────────────

async function optimizeCover(file) {
  const srcPath = path.join(COVERS, file);
  const ext = extname(file);
  const base = path.basename(file, ext);
  const webpPath = path.join(COVERS, `${base}.webp`);
  const mp4Path = path.join(COVERS, `${base}.mp4`);
  const before = (await stat(srcPath)).size;

  if (ext === ".gif") {
    // GIF → WebP poster + MP4 loop via ffmpeg (sharp can't do GIF→MP4)
    const frame = path.join(COVERS, `${base}.frame.png`);
    run(`ffmpeg -y -i "${srcPath}" -vf "scale='min(${COVER_MAX_W},iw)':-2" -frames:v 1 "${frame}"`);
    await sharp(frame).webp({ quality: WEBP_QUALITY }).toFile(webpPath);
    unlinkSync(frame);

    try {
      run(`ffmpeg -y -i "${srcPath}" -vf "scale='min(${COVER_MAX_W},iw)':-2" -c:v libx264 -pix_fmt yuv420p -movflags +faststart -an -crf 24 "${mp4Path}"`);
    } catch {
      log(`  ${file}: mp4 fallback with palette`);
      try {
        run(`ffmpeg -y -i "${srcPath}" -vf "fps=15,scale='min(${COVER_MAX_W},iw)':-2,split[s0][s1];[s0]palettegen[p];[s1][p]paletteuse" -c:v libx264 -pix_fmt yuv420p -movflags +faststart -an -crf 24 "${mp4Path}"`);
      } catch {
        log(`  ${file}: mp4 failed, keeping webp only`);
      }
    }
    const webpSize = (await stat(webpPath)).size;
    const mp4Size = existsSync(mp4Path) ? (await stat(mp4Path)).size : 0;
    log(`${file}: ${humanSize(before)} → webp ${humanSize(webpSize)}${mp4Size ? ` + mp4 ${humanSize(mp4Size)}` : ""}`);
    return;
  }

  // PNG/JPG cover → WebP
  await sharp(srcPath)
    .resize({ width: COVER_MAX_W, withoutEnlargement: true })
    .webp({ quality: WEBP_QUALITY })
    .toFile(webpPath);
  const webpSize = (await stat(webpPath)).size;
  log(`${file}: ${humanSize(before)} → webp ${humanSize(webpSize)}`);
}

// ─── ASSETS ─────────────────────────────────────────────────────────

async function optimizeAsset(file) {
  const srcPath = path.join(ASSETS, file);
  const ext = extname(file);

  // Skip if it's already a variant or webp/mp4
  if (ext !== ".png" && ext !== ".jpg" && ext !== ".jpeg") return;
  if (file.includes("_rw_")) return; // we process originals only

  const base = path.basename(file, ext);
  const before = (await stat(srcPath)).size;

  // Remove placeholder variants (176B files are broken exports)
  for (const w of [...ASSET_WIDTHS, 3840]) {
    const variantPatterns = [
      path.join(ASSETS, `${base}_rw_${w}.png`),
      path.join(ASSETS, `${base}_rw_${w}.jpg`),
      path.join(ASSETS, `${base}_rw_${w}.jpeg`),
    ];
    for (const v of variantPatterns) {
      if (existsSync(v)) {
        const vSize = (await stat(v)).size;
        if (vSize < PLACEHOLDER_SIZE) {
          unlinkSync(v);
          log(`  removed placeholder: ${path.basename(v)} (${humanSize(vSize)})`);
        }
      }
    }
  }

  // Generate WebP at needed widths
  for (const w of ASSET_WIDTHS) {
    const webpPath = path.join(ASSETS, `${base}_rw_${w}.webp`);
    if (existsSync(webpPath)) continue;

    const img = sharp(srcPath);
    const meta = await img.metadata();
    const targetW = Math.min(w, meta.width);
    await img
      .resize({ width: targetW, withoutEnlargement: true })
      .webp({ quality: WEBP_QUALITY })
      .toFile(webpPath);
  }

  // Remove old PNG/JPG variants (we now have WebP)
  for (const w of [...ASSET_WIDTHS, 3840]) {
    for (const ext of [".png", ".jpg", ".jpeg"]) {
      const v = path.join(ASSETS, `${base}_rw_${w}${ext}`);
      if (existsSync(v)) {
        const vSize = (await stat(v)).size;
        unlinkSync(v);
        log(`  removed old variant: ${path.basename(v)} (${humanSize(vSize)})`);
      }
    }
  }

  const after = (await stat(srcPath)).size;
  log(`${file}: ${humanSize(before)} → ${humanSize(after)} (originals kept)`);
}

// ─── JSON PATH UPDATES ──────────────────────────────────────────────

function updateJsonPaths() {
  let count = 0;

  function walk(dir) {
    const entries = readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(full);
        continue;
      }
      if (!entry.name.endsWith(".json")) continue;

      let text = readFileSync(full, "utf-8");
      let changed = false;

      // .gif → .webp (already done by old script, but ensure)
      if (text.includes(".gif")) {
        text = text.replace(/\.gif/g, ".webp");
        changed = true;
      }

      // .png → .webp for cover paths (the bug: old script missed these)
      if (text.includes('"cover"')) {
        text = text.replace(/"(\.\/)?\/images\/covers\/([^"]+?)\.png"/g, (match) => {
          const webpPath = path.join(COVERS, `${match.match(/\/([^\/]+)\.png/)[1]}.webp`);
          if (existsSync(webpPath)) {
            changed = true;
            return match.replace(/\.png"/, '.webp"');
          }
          return match;
        });
      }

      if (changed) {
        writeFileSync(full, text, "utf-8");
        count++;
      }
    }
  }

  walk(CONTENT);
  return count;
}

// ─── MAIN ───────────────────────────────────────────────────────────

async function main() {
  log("Starting media optimization…\n");

  // Phase 1: Covers
  log("=== Cover images ===");
  const coverFiles = await readdir(COVERS);
  for (const file of coverFiles.sort()) {
    const ext = extname(file);
    if ([".gif", ".png", ".jpg", ".jpeg"].includes(ext)) {
      await optimizeCover(file);
    }
  }

  // Phase 2: Asset images
  log("\n=== Asset images ===");
  const assetFiles = await readdir(ASSETS);
  let assetCount = 0;
  for (const file of assetFiles.sort()) {
    const ext = extname(file);
    if ([".png", ".jpg", ".jpeg"].includes(ext) && !file.includes("_rw_")) {
      await optimizeAsset(file);
      assetCount++;
    }
  }
  log(`Processed ${assetCount} original assets`);

  // Phase 3: Site images
  log("\n=== Site images ===");
  const ogJpg = path.join(SITE, "og-image.jpg");
  if (existsSync(ogJpg)) {
    const webpPath = path.join(SITE, "og-image.webp");
    if (!existsSync(webpPath)) {
      await sharp(ogJpg).resize(1200).webp({ quality: WEBP_QUALITY }).toFile(webpPath);
      log(`og-image.jpg → og-image.webp`);
    }
  }

  // Phase 4: Project video
  log("\n=== Project video ===");
  const cowboysMp4 = path.join(VIDEOS, "cowboys-vs-cyborgs-game-art.mp4");
  if (existsSync(cowboysMp4)) {
    const before = (await stat(cowboysMp4)).size;
    const tmp = cowboysMp4.replace(".mp4", ".tmp.mp4");
    run(`ffmpeg -y -i "${cowboysMp4}" -c:v libx264 -crf 24 -preset slow -movflags +faststart -an "${tmp}"`);
    unlinkSync(cowboysMp4);
    await renameAsync(tmp, cowboysMp4);
    const after = (await stat(cowboysMp4)).size;
    log(`cowboys-vs-cyborgs-game-art.mp4: ${humanSize(before)} → ${humanSize(after)}`);
  }

  // Phase 5: Update JSON paths
  log("\n=== JSON path updates ===");
  const updated = updateJsonPaths();
  log(`Updated ${updated} JSON files`);

  // Summary
  log("\n=== Summary ===");
  const totalSize = await dirSize(PUBLIC);
  log(`Total public/ size: ${humanSize(totalSize)}`);
}

async function dirSize(dir) {
  let size = 0;
  const entries = await readdir(dir, { withFileTypes: true });
  for (const entry of entries) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      size += await dirSize(full);
    } else {
      size += (await stat(full)).size;
    }
  }
  return size;
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
